dhall/src/Dhall/Binary.hs

```diff
@@ -1251,3 +1250,15 @@ instance Show DecodingFailure where
         <>  "↳ 0x" <> concatMap toHex (Data.ByteString.Lazy.unpack bytes) <> "\n"
       where
         toHex = Printf.printf "%02x "
+
+-- | This specialized version of 'Control.Monad.replicateM' reduces
+-- decoding timings by roughly 10%.
+replicateDecoder :: Int -> Decoder s a -> Decoder s [a]
+replicateDecoder n0 decoder = go n0
+  where
+    go n
+      | n <= 0    = pure []
+      | otherwise = do
+            x <- decoder
+            xs <- go (n - 1)
```

Comment:
I played with bangs (`!`) on `x` and `xs`, but this was the fastest version for my benchmark.